### PR TITLE
Example advanced forms

### PR DESCRIPTION
**Pull request by eric-burel**
HI,

I set up an example for advanced forms. This is not working yet, the custom controls don't show up for obscure reasons, but the codebase is there.

I also provided some very basic controls for objects and database item selection. If we manage to write cleaner version, I think including them into the core libs would be cool, as they would allow to handle Objects, Arrays and foreign keys out of the box in the SmartForm.

This example is related to this issue:
 [https://github.com/VulcanJS/Vulcan/issues/1863](https://github.com/VulcanJS/Vulcan/issues/1863)

### CODE REVIEW

File: packages/example-forms/lib/components/HomePage.jsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    state = {}
+    render() {
+        return (
+            <div>
+                <Helmet>
+                    <link
+                        name="bootstrap"
+                        rel="stylesheet"
+                        type="text/css"
+                        href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-alpha.5/css/bootstrap.min.css"
```
Comment (vincro) [inline packages/example-forms/lib/components/HomePage.jsx]: Am I right in thinking Bootstrap released v4 now, so no alpha required?

File: packages/example-forms/lib/components/ObjectForm.jsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    submitUpdate(){
+        const { name } = this.props // field name
+        const { value } = this.state
+        this.context.updateCurrentValues({
+            // we only store the id, Meteor then fetch the actual data
+            [name]: value
+        })
+    }
+    updateValue(fieldName, fieldValue) {
```
Comment (vincro) [inline packages/example-forms/lib/components/ObjectForm.jsx]: Inconsistent spacing, may also be worth adding some comments "a la jsdoc"

**Comment by vincro** [comment]
Interesting to see, and to be able to compare a more holistic solution to how I myself also set up ArrayFields =)

**Comment by SachaG** [comment]
Oh nice. Could we also find a way to fit in field-specific data loading? Maybe for the `customerId` field?

http://docs.vulcanjs.org/forms.html#Field-Specific-Data-Loading

File: packages/example-forms/lib/components/DataSelectForm.jsx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    render() {
+        const { label, fragmentName, ...otherProps } = this.props
+        const options = {
+            collection: this.props.collection,
+            fragment: fragmentName ? getFragment(fragmentName) : undefined,
+            ...this.props.options
+        }
+
+
+        const SelectInputWithData = withList(options)(SelectInput);
```
Comment (SachaG) [inline packages/example-forms/lib/components/DataSelectForm.jsx]: It's recommended not to use HoCs inside `render`: https://reactjs.org/docs/higher-order-components.html#dont-use-hocs-inside-the-render-method

Maybe you can do what I do here? https://github.com/VulcanJS/Vulcan/blob/devel/packages/vulcan-forms/lib/components/FormWrapper.jsx#L53

Although maybe we can get rid of `withList` altogether, depending on our needs.

File: packages/example-forms/lib/modules/customers/schema.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            document.userId = currentUser._id
+        },
+    },
+    state: {
+        type: String,
+        allowedValues: stateAllowedValues,
+        default: 'prospect',
+        optional: false,
+        searchable: true,
+        form: {
```
Comment (SachaG) [inline packages/example-forms/lib/modules/customers/schema.js]: Using `form` is more or less deprecated, I think it's better to put `options` directly on the root of the field. For passing props to the field specifically there is `fieldProperties`.

File: packages/example-forms/lib/modules/customers/schema.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        viewableBy: ['guests'],
+        editableBy: ['guests'],
+        insertableBy: ['guests'],
+    },
+    'addresses.$': addressFormSchema
+}
+
+// fields with $ must be ignored for the moment otherwise 
+// Vulcan will try to generate gql for them and fail
+schema['addresses.$'].viewableBy = undefined
```
Comment (SachaG) [inline packages/example-forms/lib/modules/customers/schema.js]: I'll need to test this out myself but we shouldn't need to do that.

**Comment by eric-burel** [comment]
Hi,

You can pass a `fragmentName` if you rather need a custom fragment to fill the `DataSelectForm`, does it match your need ?
I fixed the schemas but when trying to send the form I got an error `{"id":"app.operation_not_allowed","value":"check"}`, I am not logged in but the `guests` group is allowed to create objects.

Also, as I pointed out on Slack I can't load the components using `Components.ObjectForm` in the schema, `Components === {}` on the server though the components are correctly imported.

Thanks for the feedback, I'll try to improve the example when I have a bit more spare times, I may have another project with Vulcan in the weeks to come that is mostly enhanced forms.

**Comment by SachaG** [comment]
Now that I've had a better look at the code, we can drastically simplify `DataSelectForm` using field-level data loading since we're loading all customers anyway.

On the other hand, doing selective loading of customers based on the user's input would be a great feature to have. But it might make the example a bit too complex…?

File: packages/example-forms/lib/modules/customers/schema.js (hunk 4 above)
Comment (vincro) [reply packages/example-forms/lib/modules/customers/schema.js]: @sacha is that a SmartForm thing or SimpleSchema thing?

Comment (SachaG) [reply packages/example-forms/lib/modules/customers/schema.js]: A SmartForm thing, SimpleSchema doesn't know/care about forms.

**Comment by eric-burel** [comment]
Hi,
I think [this commit](https://github.com/VulcanJS/Vulcan/commit/bf9ebaf57c8286ff4ffe7003c6060da3e59d0f2d) will allow document creation in the example. Guests were not allowed to create new document using the default mutations, even with the correct permissions, this commit fixes it.

To my mind the `DataSelectForm` is indeed meant for lazy loading, because you can't tell the collection size at first, I just did not set up because well, I am the lazy one. React-select seems to have some features to make this easier.
I think it's okay even if it makes the example a bit complex, as long as improvements concern best practices or issues that you encounter as soon as you're coding a real life app like loading data only on demand.

**Pull request closed without merging**